components/ai_chat/resources/page/state/ai_chat_active_chat_provider.tsx

```diff
@@ -0,0 +1,104 @@
+
+type Children = (details: SelectedChatDetails) => JSX.Element
+
+const updateSelectedConversation = (selectedId: string | undefined) => {
+  window.location.href = `/${selectedId ?? ''}`
+}
+
+export function ActiveChatProviderFromUrl(props: {
+  children: Children
+}) {
+  // Register the empty route, so we don't reload the page when navigating to '/'
+  useRoute('/')
+
+  const selectedConversationId = useRoute(`/{chatId}`)?.chatId
+  return <ActiveChatProvider selectedConversationId={selectedConversationId} updateSelectedConversationId={updateSelectedConversation}>
+    {props.children}
+  </ActiveChatProvider>
+}
+
+function ActiveChatProvider({ children, selectedConversationId, updateSelectedConversationId }: {
+  selectedConversationId: string | undefined
+  updateSelectedConversationId: (selectedId: string | undefined) => void,
+  children: Children
+}) {
+  const { visibleConversations } = useAIChat()
+  const [conversationAPI, setConversationAPI] =
+    React.useState<Pick<SelectedChatDetails, 'callbackRouter' | 'conversationHandler'>>()
+
+  const details = React.useMemo(() => ({
+    ...conversationAPI,
+    selectedConversationId,
+    updateSelectedConversationId,
+  }), [selectedConversationId, updateSelectedConversationId, conversationAPI])
+
+  React.useEffect(() => {
+    // Handle creating a new conversation
+    if (!selectedConversationId) {
+      setConversationAPI(API.newConversation())
+      return
+    }
+
+    // Select a specific conversation
+    setConversationAPI(API.bindConversation(selectedConversationId === tabAssociatedChatId
+      ? undefined
+      : selectedConversationId))
+
+    // The default conversation changes as the associated tab navigates, so
+    // listen for changes.
+    if (selectedConversationId === tabAssociatedChatId) {
+      const onNewDefaultConversationListenerId =
+        getAPI().UIObserver.onNewDefaultConversation.addListener(() => {
+          setConversationAPI(API.bindConversation(undefined))
+        })
+
+      return () => {
+        getAPI().UIObserver.removeListener(onNewDefaultConversationListenerId)
+      }
+    }
+
+    // Satisfy linter
+    return undefined
+  }, [selectedConversationId])
+
+  // Clean up bindings when not used anymore
+  React.useEffect(() => {
+    return () => {
+      conversationAPI?.callbackRouter.$.close()
+      conversationAPI?.conversationHandler.$.close()
+    }
+  }, [conversationAPI])
+
+  // Handle the case where a non-existent chat has been selected:
+  React.useEffect(() => {
+    // Special case the default conversation - it gets treated specially as
+    // the chat is rebound as the tab navigates.
+    if (selectedConversationId === tabAssociatedChatId) return
+
+    if (selectedConversationId && !visibleConversations.find(c => c.uuid === selectedConversationId)) {
+      updateSelectedConversationId(undefined)
+    }
+  }, [visibleConversations, selectedConversationId])
+
+  return conversationAPI
+    ? children(details as any)
```

Comment:
What's the reason for this slightly non-standard way of passing the data? Why not just use context? Is it to protect it from descendent components and make sure only the direct child gets access to the data? Either way, I think we need a comment at the top of this explaining its use.

Reply:
This is the render props pattern which is pretty standard in React - basically, setting up another context seemed like overkill here, especially because only the `ConversationContext` is concerned with the data it provides.

Reply:
https://legacy.reactjs.org/docs/render-props.html

Reply:
(added the comment)

Reply:
ok I think I see why you prefer this over a custom hook, just about...